File: tests/xcalablemp/NOT-IMPLEMENTED/59/testp111.f90
! testp111.f
! reflect指示文、1次元分散、width、asyncあり

      program main
      include 'xmp_lib.h'
      integer,parameter:: N=1000
!$xmp nodes p(*)
!$xmp template t(N)
!$xmp distribute t(block) onto p
      integer a(N)
      real*8  b(N)
      real*4  c(N)
!$xmp align a(i) with t(i)
!$xmp align b(i) with t(i)
!$xmp align c(i) with t(i)
!$xmp shadow a(1)
!$xmp shadow b(2)
!$xmp shadow c(3)
      integer w
      character(len=3) result

      if(mod(N,xmp_num_nodes()) .eq. 0) then
         w = N/xmp_num_nodes()
      else
         w = N/xmp_num_nodes()+1
      endif
      
      result = 'OK '
      k = 1
!$xmp loop on t(i)
      do i=1, N
         a(i) = k
         b(i) = dble(k)
         c(i) = real(k)
      enddo
      
!$xmp reflect (a) async(1)
!$xmp reflect (b) async(2)
!$xmp reflect (c) async(3)

!$xmp wait_async(1)
!$xmp loop on t(i)
      do i=2, N-1
         do ii=i-1,i+1
            if(a(ii) .ne. k) result = 'NG1'
         enddo
      enddo
!$xmp wait_async(2)
!$xmp loop on t(i)
      do i=3, N-2
         do ii=i-2,i+2
            if(b(ii) .ne. dble(k)) result = 'NG2'
         enddo
      enddo
!$xmp wait_async(3)
!$xmp loop on t(i)
      do i=4, N-3
         do ii=i-3,i+3
            if(c(ii) .ne. real(k)) result = 'NG3'
         enddo
      enddo
      
      k = 2
!$xmp loop on t(i)
      do i=1, N
         a(i) = k
         b(i) = dble(k)
         c(i) = real(k)
      enddo
      
!$xmp reflect (a) async(1)
!$xmp reflect (b) width(2) async(2)
!$xmp reflect (c) width(2) async(3)
!$xmp wait_async(1)

!$xmp loop on t(i)
      do i=2, N-1
         do ii=i-1,i+1
            if(a(ii) .ne. k) result = 'NG4'
         enddo
      enddo
!$xmp wait_async(2)
!$xmp loop on t(i)
      do i=3, N-2
         do ii=i-2,i+2
            if(b(ii) .ne. dble(k)) result = 'NG5'
         enddo
      enddo
!$xmp wait_async(3)
!$xmp loop on t(i)
      do i=4, N-3
         do ii=i-2,i+2
            if(c(ii) .ne. real(k)) result = 'NG6'
         enddo
         if(i.eq.4 .or. i.eq.w+1) then
            if(c(i-3) .ne. 1.0) result = 'NG7'
         endif
         if(i.eq.N-3 .or. i.eq.w*xmp_node_num()) then
            if(c(i+3) .ne. 1.0) result = 'NG8'
         endif
      enddo

      print *, xmp_node_num(), 'testp111.f ', result

      end

File: tests/xcalablemp/NOT-IMPLEMENTED/59/testp111.c
/*testp111.c*/
/*reflect指示文,1次元分散,width,asyncあり*/
#include<mpi.h>
#include<xmp.h>
#include<stdio.h>   
#pragma xmp nodes p(*)
static const int N=1000;
#pragma xmp template t(0:N-1)
#pragma xmp distribute t(block) onto p
int a[N];
double b[N];
float c[N];
#pragma xmp align a[i] with t(i)
#pragma xmp align b[i] with t(i)
#pragma xmp align c[i] with t(i)
#pragma xmp shadow a[1]
#pragma xmp shadow b[2]
#pragma xmp shadow c[3]
int w,i,ii,k;
char *result;
int main(void){
   if(N%xmp_num_nodes() == 0){
      w = N/xmp_num_nodes();
   }else{
      w = N/xmp_num_nodes()+1;
   }
   
   result = "OK";
   k = 1;
#pragma xmp loop on t(i)
   for(i=0;i<N;i++){
      a[i] = k;
      b[i] = (double)k;
      c[i] = (float)k;
   }
#pragma xmp reflect (a) async(1)
#pragma xmp reflect (b) async(2)
#pragma xmp reflect (c) async(3)
#pragma xmp wait_async(1)
#pragma xmp loop on t(i)
   for(i=2;i<N;i++){
      for(ii=i-1;ii<i+2;ii++){
         if(a[ii] != k) result = "NG1";
      }
   }
#pragma xmp wait_async(2)
#pragma xmp loop on t(i)
   for(i=3;i<N-1;i++){
      for(ii=i-2;ii<i+3;ii++){
         if(b[ii] != (double)k) result = "NG2";
      }
   }
#pragma xmp wait_async(3)
#pragma xmp loop on t(i)
   for(i=4;i<N-2;i++){
      for(ii=i-3;ii<i+4;ii++){
         if(c[ii] != (float)k) result = "NG3";
      }
   }
   k = 2;
#pragma xmp loop on t(i)
   for(i=0;i<N;i++){
      a[i] = k;
      b[i] = (double)k;
      c[i] = (float)k;
   }
#pragma xmp reflect (a) async(1)
#pragma xmp reflect (b) width(2) async(2)
#pragma xmp reflect (c) width(2) async(3)
#pragma xmp wait_async(1)

#pragma xmp loop on t(i)
   for(i=2;i<N;i++){
      for(ii=i-1;ii<i+2;ii++){
         if(a[ii] != k) result = "NG4";
      }
   }
#pragma xmp wait_async(2)
#pragma xmp loop on t(i)
   for(i=3;i<N-1;i++){
      for(ii=i-2;ii<i+3;ii++){
         if(b[ii] != (double)k) result = "NG5";
      }
   }
#pragma xmp wait_async(3)
#pragma xmp loop on t(i)
   for(i=4;i<N-2;i++){
      for(ii=i-3;ii<i+4;ii++){
         if(c[ii] != (float)k) result = "NG6";
      }
      if(i==4||i==w+1){
         if(c[i-3] != 1.0) result = "NG7";
      }
      if(i==N-3||i==w*xmp_node_num()){
         if(c[i+3]!=1.0) result = "NG8";
      }
   } 
   printf("%d %s %s\n",xmp_node_num(),"testp111.c",result); 

   return 0;
}
